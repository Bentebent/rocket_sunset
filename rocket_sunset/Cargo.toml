[package]
name = "rocket_sunset"
version = "0.1.0"
edition = "2021"
publish = true

[lib]
crate-type = ["cdylib", "rlib"]
name = "rocket_sunset"
path = "src/lib.rs"

[dev-dependencies]
trybuild = "1.0.41"

[dependencies]
rocket = { version = "0.5.1", features = ["json"] }
sunset_macro = { version = "0.1.0", path = "../sunset_macro" }
