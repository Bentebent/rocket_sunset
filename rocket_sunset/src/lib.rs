use rocket::{
    self,
    http::Header,
    response::{
        self,
        Responder,
    },
    Request,
};
pub use rocket_sunset_macro::deprecation;

pub struct DeprecatedResponder<T: for<'r> Responder<'r, 'static>> {
    pub inner: T,
    pub deprecation: Header<'static>,
    pub link: Option<Header<'static>>,
    pub sunset: Option<Header<'static>>,
}

impl<T: for<'r> Responder<'r, 'static>> DeprecatedResponder<T> {
    pub fn new(inner: T, deprecated_timestamp: &str, link: Option<&str>, sunset: Option<&str>) -> Self {
        DeprecatedResponder {
            inner,
            deprecation: Header::new("deprecation", format!("@{}", deprecated_timestamp)),
            link: link.map(|l| Header::new("link", format!(r#"<{}>; rel="deprecation"; type="text/html""#, l))),
            sunset: sunset.map(|s| Header::new("sunset", s.to_owned())),
        }
    }
}

impl<'r, 'o: 'r, T> Responder<'r, 'o> for DeprecatedResponder<T>
where
    T: for<'any> Responder<'any, 'static>,
{
    fn respond_to(self, req: &'r Request<'_>) -> response::Result<'o> {
        self.inner.respond_to(req).map(|mut response| {
            response.set_header(self.deprecation);
            if let Some(header) = self.link {
                response.set_header(header);
            }
            if let Some(header) = self.sunset {
                response.set_header(header);
            }
            response
        })
    }
}
