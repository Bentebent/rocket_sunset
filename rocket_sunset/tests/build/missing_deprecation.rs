use rocket::{
    get,
    routes,
};
use rocket_sunset::{
    deprecation,
    DeprecatedResponder,
};

#[get("/")]
#[deprecation(sunset = "2024-01-01T00:00:00Z")]
pub fn index() -> &'static str {
    "Hello, world!"
}

fn main() {
    let _ = rocket::build().mount("/", routes![index]);
}
