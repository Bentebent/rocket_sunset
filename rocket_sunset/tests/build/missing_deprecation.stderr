error: Missing or invalid deprecation timestamp, expected string literal
  --> tests/build/missing_deprecation.rs:11:15
   |
11 | #[deprecation(sunset = "2024-01-01T00:00:00Z")]
   |               ^^^^^^

warning: unused import: `DeprecatedResponder`
 --> tests/build/missing_deprecation.rs:7:5
  |
7 |     DeprecatedResponder,
  |     ^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error[E0423]: expected function, found struct `index`
  --> tests/build/missing_deprecation.rs:12:19
   |
10 | #[get("/")]
   | -----------
   | |
   | `index` defined here
   | in this procedural macro expansion
11 | #[deprecation(sunset = "2024-01-01T00:00:00Z")]
12 | pub fn index() -> &'static str {
   |                   ^ help: use struct literal syntax instead: `index {}`
   |
   = note: this error originates in the attribute macro `get` (in Nightly builds, run with -Z macro-backtrace for more info)
